### ISSUE TITLE
: unrecognized arguments: model_def:MyTrial

### ISSUE THREAD

**Issue by caiduoduo12138** [determined-ai/determined#4817]
[2022-08-18 09:40:47] [14ac6e30] harness.py: error: unrecognized arguments: model_def:MyTrial
<error> [2022-08-18 09:40:47] [14ac6e30] resources failed with non-zero exit code: container failed with non-zero exit code: 2 (exit code 2)

detailed:
[2022-08-18 09:40:44]
[14ac6e30] [23] root: detected 1 gpus <warning> [2022-08-18 09:40:45]
[14ac6e30] [23] root: process [Not Found] with pid 208068 is using 3647 MiB memory of the GPU with uuid GPU-3589f80e-875f-69b6-64ab-45f8e1f3e841. This process is not related to Determined tasks but may interfere with tasks' ability to use the full GPU. <info> [2022-08-18 09:40:45]
[14ac6e30] [29] root: New trial runner in (container 14ac6e30-1449-4595-84b9-1529fb3f3856) on agent node5: {"bind_mounts": [], "checkpoint_policy": "best", "checkpoint_storage": {"host_path": "/root/.local/share/determined", "propagation": "rprivate", "save_experiment_best": 0, "save_trial_best": 1, "save_trial_latest": 1, "storage_path": null, "type": "shared_fs"}, "data_layer": {"container_storage_path": null, "host_storage_path": null, "type": "shared_fs"}, "data": {}, "debug": false, "description": null, "entrypoint": "model_def:MyTrial", "environment": {"image": {"cpu": "determinedai/environments:py-3.8-pytorch-1.10-tf-2.8-cpu-9119094", "cuda": "determinedai/environments:cuda-11.3-pytorch-1.10-tf-2.8-gpu-9119094", "rocm": "determinedai/environments:rocm-4.2-pytorch-1.9-tf-2.5-rocm-9119094"}, "environment_variables": {"cpu": [], "cuda": [], "rocm": []}, "ports": {"trial": 1734}, "registry_auth": null, "force_pull_image": false, "pod_spec": null, "slurm": [], "add_capabilities": [], "drop_capabilities": []}, "hyperparameters": {"global_batch_size": {"type": "const", "val": 4}, "learning_rate": {"type": "const", "val": 0.0001}}, "labels": [], "max_restarts": 0, "min_checkpoint_period": {"batches": 0}, "min_validation_period": {"epochs": 1}, "name": "MyTrial", "optimizations": {"aggregation_frequency": 1, "average_aggregated_gradients": true, "average_training_metrics": true, "gradient_compression": false, "grad_updates_size_file": null, "mixed_precision": "O0", "tensor_fusion_threshold": 64, "tensor_fusion_cycle_time": 5, "auto_tune_tensor_fusion": false}, "perform_initial_validation": false, "profiling": {"enabled": false, "begin_on_batch": 0, "end_after_batch": null, "sync_timings": true}, "project": "", "records_per_epoch": 50000, "reproducibility": {"experiment_seed": 1660815632}, "resources": {"max_slots": null, "slots_per_trial": 1, "weight": 1, "native_parallel": false, "shm_size": null, "agent_label": "", "resource_pool": "default", "priority": null, "devices": []}, "scheduling_unit": 100, "searcher": {"max_length": {"epochs": 32}, "metric": "validation_error", "name": "single", "smaller_is_better": true, "source_checkpoint_uuid": null, "source_trial_id": null}, "workspace": ""} <info> [2022-08-18 09:40:45]
[14ac6e30] [29] root: Validating checkpoint storage ... <info> [2022-08-18 09:40:45]
[14ac6e30] [29] root: Launching: ['python3', '-m', 'determined.launch.horovod', '--autohorovod', '--trial', 'model_def:MyTrial'] <info> [2022-08-18 09:40:46]
[14ac6e30] [33] root: Loading Trial implementation with entrypoint model_def:MyTrial. <none> [2022-08-18 09:40:47]
[14ac6e30] usage: harness.py [-h] [--lr LR] [--lr1 LR1] [--lr2 LR2] [--resume] [--bs BS] <none> [2022-08-18 09:40:47]
[14ac6e30]                   [--wd WD] [--schedule SCHEDULE [SCHEDULE ...]] <none> [2022-08-18 09:40:47]
[14ac6e30]                   [--gamma GAMMA] [--alg ALG] [--epochs EPOCHS] <none> [2022-08-18 09:40:47]
[14ac6e30]                   [--state_epoch STATE_EPOCH] [--path PATH] [--model MODEL] <none> [2022-08-18 09:40:47]
[14ac6e30]                   [--lr-type {cosine,multistep,bigcycle,smallcycle}] <none> [2022-08-18 09:40:47]
[14ac6e30]                   [--data {cifar10,cifar100,imagenet}] [--arch ARCH] [--s S] <none> [2022-08-18 09:40:47]
[14ac6e30]                   [--i I] [--num NUM] <none> [2022-08-18 09:40:47]
[14ac6e30] harness.py: error: unrecognized arguments: model_def:MyTrial 

my custom object is in model_def.py which named MyTrial
Why it can not be found?

**Comment by caiduoduo12138** [comment]
I  can run the scipt by set the entrypoint: python train.py
My config is below:
name: MyTrial
hyperparameters:
  learning_rate: 1.0e-4
  global_batch_size: 4
records_per_epoch: 50000
searcher:
  name: single
  metric: validation_error
  max_length:
    epochs: 32
min_validation_period:
  epochs: 1
max_restarts: 0
entrypoint: model_def:MyTrial

**Comment by caiduoduo12138** [comment]
My model_def.py is below: 
"""
CNN on Cifar10 from Keras example:
https://github.com/fchollet/keras/blob/master/examples/cifar10_cnn.py
"""
import filelock
import os
import tempfile
from typing import Any, Dict, Sequence, Tuple, Union, cast

import torch
import torchvision
from torch import nn
from torchvision import transforms

from determined.pytorch import DataLoader, PyTorchTrial, PyTorchTrialContext

import models.resnet
from models.SGD_GC import SGD_GC
from datasets.all import AllDataset



TorchData = Union[Dict[str, torch.Tensor], Sequence[torch.Tensor], torch.Tensor]


def accuracy_rate(predictions: torch.Tensor, labels: torch.Tensor) -> float:
    """Return the accuracy rate based on dense predictions and sparse labels."""
    assert len(predictions) == len(labels), "Predictions and labels must have the same length."
    assert len(labels.shape) == 1, "Labels must be a column vector."

    return (  # type: ignore
        float((predictions.argmax(1) == labels.to(torch.long)).sum()) / predictions.shape[0]
    )


class MyTrial(PyTorchTrial):
    def __init__(self, context: PyTorchTrialContext) -> None:
        self.context = context

        model = models.resnet.newresnet50(pretrained=True)
        self.model = self.context.wrap_model(model)

        self.optimizer = self.context.wrap_optimizer(
            SGD_GC(self.model.parameters(), lr=self.context.get_hparam("learning_rate"), momentum=0.9, weight_decay=5e-4)
        )

    def train_batch(
        self, batch: TorchData, epoch_idx: int, batch_idx: int
    ) -> Dict[str, torch.Tensor]:
        batch = cast(Tuple[torch.Tensor, torch.Tensor], batch)
        data, labels = batch

        output = self.model(data)
        loss = torch.nn.functional.cross_entropy(output, labels)
        accuracy = accuracy_rate(output, labels)

        self.context.backward(loss)
        self.context.step_optimizer(self.optimizer)
        return {"loss": loss, "train_error": 1.0 - accuracy, "train_accuracy": accuracy}

    def evaluate_batch(self, batch: TorchData) -> Dict[str, Any]:
        """
        Calculate validation metrics for a batch and return them as a dictionary.
        This method is not necessary if the user defines evaluate_full_dataset().
        """
        batch = cast(Tuple[torch.Tensor, torch.Tensor], batch)
        data, labels = batch

        output = self.model(data)
        accuracy = accuracy_rate(output, labels)
        return {"validation_accuracy": accuracy, "validation_error": 1.0 - accuracy}

    def _download_dataset(self, train: bool) -> Any:
        transform = transforms.Compose(
            [transforms.ToTensor(), transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))]
        )
        # Use a file lock so that workers on the same node attempt the download one at a time.
        # The first worker will actually perform the download, while the subsequent workers will
        # see that the dataset is downloaded and skip.
        with filelock.FileLock(os.path.join(self.download_directory, "lock")):
            return torchvision.datasets.CIFAR10(
                root=self.download_directory, train=train, download=True, transform=transform
            )

    def build_training_data_loader(self) -> Any:
        train_transform = transforms.Compose([
            transforms.Resize((512, 384)),
            transforms.ToTensor(),
            transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225]),
        ])
        train_data = AllDataset('/root/test_dist/images/', '/root/test_dist/default.txt', transform=train_transform, mytransform=False)
        train_loader = DataLoader(dataset=train_data, batch_size=self.context.get_per_slot_batch_size(), shuffle=True)

        # trainset = self._download_dataset(train=True)
        # return DataLoader(trainset, batch_size=self.context.get_per_slot_batch_size())
        return train_loader

    def build_validation_data_loader(self) -> Any:
        val_transform = transforms.Compose([
            transforms.Resize((512, 384)),
            transforms.ToTensor(),
            transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225]),
        ])
        val_data = AllDataset('/root/test_dist/images/', '/root/test_dist/default.txt', transform=val_transform, mytransform=False)
        val_loader = DataLoader(dataset=val_data, batch_size=self.context.get_per_slot_batch_size(), shuffle=True)
        return val_loader

**Comment by ioga** [comment]
1. Looking at the error message: 
```
[14ac6e30] usage: harness.py [-h] [--lr LR] [--lr1 LR1] [--lr2 LR2] [--resume] [--bs BS] [2022-08-18 09:40:47]
[14ac6e30] [--wd WD] [--schedule SCHEDULE [SCHEDULE ...]] [2022-08-18 09:40:47]
[14ac6e30] [--gamma GAMMA] [--alg ALG] [--epochs EPOCHS] [2022-08-18 09:40:47]
[14ac6e30] [--state_epoch STATE_EPOCH] [--path PATH] [--model MODEL] [2022-08-18 09:40:47]
[14ac6e30] [--lr-type {cosine,multistep,bigcycle,smallcycle}] [2022-08-18 09:40:47]
[14ac6e30] [--data {cifar10,cifar100,imagenet}] [--arch ARCH] [--s S] [2022-08-18 09:40:47]
[14ac6e30] [--i I] [--num NUM] [2022-08-18 09:40:47]
```
these flags don't look familiar to me. do you happen to have a file called "harness.py" with the same flag in your model that may be conflicting and getting imported by mistake?

2. it seems like your code is based on our `cifar10_pytorch` example. Does it work for you? you can do an ablation study to see what change you made to the example has caused it to start failing.

**Comment by caiduoduo12138** [comment]
Thank you for your advice!

**Comment by caiduoduo12138** [comment]
Thank you for your advice!

**Comment by caiduoduo12138** [comment]
> Thank you for your advice!
